### ISSUE TITLE
scikit-learn lecture

### ISSUE THREAD

**Issue by john-science** [theJollySin/python_for_scientists#69]
This seems like a big library. But okay, let's learn it.

## What is it?

The [scikit-learn main page](https://scikit-learn.org/stable/index.html) says the it deals with three kinds of learning:

* Classification
* Regression
* Clustering

## The API

Looking at [the api](https://scikit-learn.org/stable/modules/classes.html), it has 19 algorithms modules and five modules that look like data input stuff:

### Algorithms and Learning Techniques

* **cluster** - common clustering algorithms
* **covariance** - covariance algorithms
* **decomposition** - various matrix decomposition algorithms
* **ensemble** - "ensemble-based" methods for classification and regression
* **feature_selection** - common statistical algorithms for selecting features in data
* **gaussian_process** - Gaussian regression and classification
* **linear_model** - linear regression / gradient ascent
* **semi_supervised** - special class of learning algorithms, never heard of them, neat
* **svm** - support vector machines
* **tree** - decision trees for classification and regression
* **mixture** - Bayesian / Gaussian Mixture modeling algorithms
* **model_selection** - I *guess* a few statistical algorithms to make selections
* **neighbors** - "k-nearest" neighbors
* **neural_network** - neural networks

### Data Tools

* **feature_extraction** - tools to read data files
* **preprocessing** - common statistics tools for preprocessing data
* **manifold** - "data embedding techniques"
* **metrics** - score functions, performance / pairwise metrics
* **datasets** - example datasets

## How do we learn this?

Learn by doing.

The website is full of wonderful practical examples, with data sets and everything. Pick an example from each of the Algorithm / Techniques modules above, and do them in turn.

As time goes on, you will see how data is input, how to clean data, and how to use the tools in the library.  And once you see enough parallels of that process, assuming the library is well designed, you should just be able to grab the library to solve any problem. Assuming you understand the underlying algorithm. And you think you have a problem that is susceptible to a standard / canned approach.

**Comment by john-science** [comment]
This lecture would be a big undertaking.  The `scikit-learn` library is just a grab bag of various learning algorithms and tools.

**Comment by john-science** [comment]
I'm not sure this library is learnable in a half-hour class. It doesn't have enough structure or organization. It's just a hodge-podge pile of algorithms, without a high enough level API to really "cover" the whole library in one short explanation. 

I will probably cover this library, but not in the format of this class.
